### ISSUE TITLE
Re-enable feature validation in sklearn predict proba method by default

### ISSUE THREAD

**Issue by stephenpardy** [dmlc/xgboost#6158]
XGBoost version 1.2 changed the default value of `validate_features` from `True` to `False` in the scikit-learn `predict_proba` method (https://github.com/dmlc/xgboost/pull/5953/files) with limited discussion. This exposes the user to the risk that they will silently receive the wrong predictions using this version of XGBoost - especially since most users will expect this function to raise a ValueError by default.

The most likely case where this would pose a problem is when you fit a model using some subset of columns from a file and then read in that data without first properly ordering that data. This can easily happen because Pandas does not respect the ordering that you pass in to `usecols`. Here is a minimum example:

```
import numpy as np
import pandas as pd
import xgboost

data = pd.DataFrame(
    data=np.random.normal(loc=range(10), size=(1000, 10)),
    columns=[f"x{i}" for i in range(10)],
)

target = data.sum(axis=1)
data["TARGET"] = target
data.to_csv("dummy_data.csv", index=False)

model = xgboost.XGBRegressor()
model.fit(
    # note that we build this model with features pulled out-of-order from our dataframe
    data[["x2", "x5", "x3", "x1", "x9"]],
    data["TARGET"]
)

orig_preds = model.predict(data[["x2", "x5", "x3", "x1", "x9"]])

new_data = pd.read_csv(
    "dummy_data.csv",
    # load only the features that were used in the original model
    usecols=model.get_booster().feature_names
)

np.allclose(orig_preds, model.predict(new_data, validate_features=False))  # False

np.allclose(orig_preds, model.predict(
    new_data[model.get_booster().feature_names], validate_features=True
))  # True
```

**Comment by trivialfis** [comment]
Em, now we have a conflicting situation.

**Comment by hcho3** [comment]
@stephenpardy The change was to fix #5887, where scikit-learn takes in a Pandas dataframe but outputs a NumPy array, leading to the loss of feature names. Also see scikit-learn/scikit-learn#5523. Do you have any suggestion about how we can enable `GridSearchCV` or `CalibratedClassifierCV` with feature validation?

**Comment by stephenpardy** [comment]
This does seem to be a rock and a hard place situation. On sklearn version 0.22 at least the `GridSearchCV` is working fine with `validate_features`. If this is only effecting `CalibratedClassifierCV`, is it reasonable to make users monkey-patch `predict_proba` to make it work with XGBoost until there is a fix from SKLearn.

This works fine with my setup:
```
kfolds = StratifiedKFold(3)

model = xgboost.XGBClassifier()

params = {}

clf = GridSearchCV(
    model,
    params,
    cv=kfolds.split(x_train, y_train),
    scoring='roc_auc', return_train_score=True
)

clf.fit(x_train, y_train)

model = clf.best_estimator_

old_predictor = model.predict_proba
def predict_proba_no_validate(X, y=None):
    return old_predictor(X, y, validate_features=False)

model.predict_proba = predict_proba_no_validate

clf_isotonic = CalibratedClassifierCV(model, cv='prefit', method='isotonic')
clf_isotonic.fit(x_valid, y_valid)
```

If the SKLearn `CalibratedClassifierCV` could take an additional set of keyword arguments that passed on to the `predict_proba` method that would also be a cleaner temporary solution to this issue.

Finally, if none of the above solutions are possible I think that at a minimum it would be good to have a warning raised to the user (instead of the old default Exception) so that this doesn't fail silently.

**Comment by trivialfis** [comment]
I want to remove the feature name generation for arrays.  This way there's no feature name validation for arrays but works with dataframe consistently.  From my perspective, XGBoost is not supposed to invent these feature names except for plotting.
